### PR TITLE
Editing a topic

### PR DESCRIPTION
**Pull request by Mentessi**
This introduces repetition of the code used in my form - I need to separate this out into a partial (I think?) - I'll do this after you've reviewed my other PRs as I dont want the merging to get too awkward (I'm not 100% sure how complicated it would be, just erring on the side of caution)

### CODE REVIEW

File: app/controllers/topics_controller.rb [hunk 1]
```diff
@@ -16,6 +16,24 @@ def create
...
+  def show
+    @topic = Topic.find_by_id(params[:id])
+
+    if @topic.nil?
+      redirect_to topics_path
+    end
+  end
+
+  def edit
+    @topic = Topic.find_by_id(params[:id])
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: Just use the `find` method which defaults to using the id.
```
@topic = Topic.find(params[:id])
```

Comment (Neilos) [reply app/controllers/topics_controller.rb]: Hmmm... 🤔 
Notice how that line keeps has repeated in four actions inside this controller. Whenever you see something repeating 3 or more times, it's usually a good time to think about DRYing up your code.

In this case, a `before_action` call (previously called a `before_filter` in earlier versions of Rails) will do the job. I'll let you look that up. You could even do it as a separate pull request if you like.

File: app/controllers/topics_controller.rb [hunk 2]
```diff
@@ -16,6 +16,24 @@ def create
...
+  end
+
+  def update
+    @topic = Topic.find_by_id(params[:id])
+    @topic.update!(topic_params)
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: So, if the user makes a mistake this controller action will just blow up and present the user with the 500 page. 
What about all the hard work you did to validate the Topic?
Take a look at your `create` action and give this another go.

File: app/controllers/topics_controller.rb [hunk 3]
```diff
@@ -16,6 +16,24 @@ def create
...
+    redirect_to topic_path(@topic.id)
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: Just `redirect_to topic_path(topic)` will do. No need to convert to an id.

File: app/controllers/topics_controller.rb [hunk 4]
```diff
@@ -16,6 +16,24 @@ def create
...
+      redirect_to topics_path
+    end
+  end
+
+  def edit
+    @topic = Topic.find_by_id(params[:id])
+  end
+
+  def update
+    @topic = Topic.find_by_id(params[:id])
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: Just use `find` method.

File: spec/system/managing_topics_spec.rb [hunk 5]
```diff
@@ -21,4 +21,22 @@
...
+    visit topics_path
+    click_on('What is your favourite colour')
+    expect(current_path).to eq topic_path(@topic.id)
+    expect(page).to have_content('Topic: 1')
+  end
+
+  scenario 'when editing a topic' do
+    visit topic_path(@topic.id)
+    click_link 'Edit Topic'
+    expect(current_path).to eq edit_topic_path(@topic.id)
```
Comment (Neilos) [inline spec/system/managing_topics_spec.rb]: Just `edit_topic_path(topic)` will do. No need for conversion to an id

Comment (Neilos) [reply spec/system/managing_topics_spec.rb]: Also, use `let` blocks in specs instead of instance variables.

File: spec/system/managing_topics_spec.rb [hunk 6]
```diff
@@ -21,4 +21,22 @@
...
+    expect(page).to have_content 'What is your favourite colour'
+    fill_in 'Name', with: 'this is my new topic'
```
Comment (Neilos) [inline spec/system/managing_topics_spec.rb]: What about trying out an error (e.g. submitting with a blank name), then you can see if you get the validation error that you expect. (Hint: it's not working at present)

File: spec/system/managing_topics_spec.rb [hunk 7]
```diff
@@ -21,4 +21,22 @@
...
+    click_button 'Update Topic'
+    expect(current_path).to eq topic_path(@topic.id)
```
Comment (Neilos) [inline spec/system/managing_topics_spec.rb]: Just `topic_path(topic)` will do. No need for conversion to an id

File: app/controllers/topics_controller.rb [hunk 8]
```diff
@@ -17,11 +20,31 @@ def create
...
+      redirect_to topic_path
+    else
+      @topic.name = Topic.find(params[:id]).name
+      render :edit
+    end
   end
 
   private
 
+  def topic_find
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: Really minor naming point (no need to change)....
There are two types of methods: methods that return values (functions) and methods that have side effects. (In fact all methods return values, but for some methods you don't care about the value returned)

My rules for naming are:
- When a method returns a value (i.e. a function) name the method after the the thing being returned (e.g. "topic", "user", "tag_collection").
- When a method has a side effect name the method in the present tense imperative (e.g. "run away", "stop what you are doing" "sit down").

In this case the method has a side effect of setting an instance variable, so I'd name it `find_topic`

File: spec/system/managing_topics_spec.rb [hunk 9]
```diff
@@ -27,4 +27,26 @@
     expect(current_path).to eq topic_path(topic)
     expect(page).to have_content('Topic:')
   end
+
+  scenario 'when deleting a topic' do
+    visit topic_path(topic)
+    # expect {click_on 'Delete Topic'}.to change(Topic, :count).by(-1)
```
Comment (Neilos) [inline spec/system/managing_topics_spec.rb]: Commented out code?

File: app/controllers/topics_controller.rb [hunk 10]
```diff
@@ -17,11 +20,31 @@ def create
...
   def show
-    @topic = Topic.find(params[:id])
+  end
+
+  def destroy
+    Topic.find(params[:id]).destroy
+    redirect_to topics_path
+  end
+
+  def edit
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: Leaving an empty method feels weird, but it's fine.

File: app/controllers/topics_controller.rb [hunk 11]
```diff
@@ -17,11 +20,31 @@ def create
...
+  end
+
+  def update
+    if @topic.update(topic_params)
+      redirect_to topic_path
+    else
+      @topic.name = Topic.find(params[:id]).name
```
Comment (Neilos) [inline app/controllers/topics_controller.rb]: You don't need this line. You want to present the topic in its edited form, so the user can see the problem.

**Comment by Mentessi** [comment]
The reason why I used - @topic.name = Topic.find(params[:id]).name - was without it the user will not know what the topic they were editing was... it's "" in @topic so it's also removed from the page... I was talking to Iain about this and he shared this with me: 

http://api.rubyonrails.org/classes/ActiveModel/Dirty.html

That looks like a solution I could use if this becomes a reoccurring issue elsewhere (or maybe you deal with this in a different way?)

File: app/controllers/topics_controller.rb (hunk 11 above)
Comment (Neilos) [reply app/controllers/topics_controller.rb]: (re your reply posted as a PR comment)

The main problem from a technical perspective is that you are doing two `find`s. That's two database queries, where only one is required.

You *could* use `ActiveModel::Dirty` as Iain suggested.
```
@topic.restore_attributes([:name])
```

...but why bother and what problem are you really solving? 

The reason the user can't identify the topic without a name is that there is no other data in the page, which means that there are no other edits to worry about and they can just navigate away from the page (e.g. go back to the topics#index page) and all is good (i.e. the name is never changed).

And so what if they enter a completely different topic name at this stage; the only thing identifying the topic is the name anyway. You don't currently have any linked tags or videos that need to match the topic name, and when you do, you'll have to worry about the issue of changing the topic name full stop (i.e. not just blanking it). Remember [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

Basically, this is probably its own feature. Whenever you add something as a nice little extra onto a feature, you introduce a whole load of complexity that isn't required, and often your quick fix doesn't fully/properly address the thing you're trying to address. T

I'd say think about what you want to happen here and why and capture it as a new story, and for this pull request, remove the line.

**Comment by Mentessi** [comment]
Dude - you are hardcore YAGI ;) .... getting out of the habit of doing things in a hacky way to get the behaviour I want rather than being happy that something works as it should and then adding a feature for the behaviour that's missing is hard (I couldn't even explain what I'm saying in a coherent sentence FFS :D ).... cheers for taking the time to write that... most useful.
